proposals/0046-astro-env.md

```diff
@@ -0,0 +1,366 @@
+- Testing static/dynamic
+- Integrations
+
+Features will be implemented incrementally using feature flags:
+
+```ts
+export default defineConfig({
+  experimental: {
+    env: {
+      schema: {},
+      // then
+      staticVariables: true,
+      // then
+      dynamicVariables: true
+    }
+  }
+})
+```
+
+# Drawbacks / integration
+
+- The implementation in terms of code size and complexity seems reasonable. It shouldn't affect too many parts of the codebase
+- The user can be implemented in userland, although more limited (eg. exposing apis to adapter is not possible without integrations inter-communication), see https://github.com/florian-lefebvre/astro-env/pull/4
+- This feature requires update across the docs as the official recommendation, although it's not breaking
+- Other parts of Astro need this, for example integrations like `@astrojs/db`
+
+# Alternatives
+
+## Typed `import.meta.env`
+
+https://github.com/florian-lefebvre/astro-env currently does [manual typing](https://docs.astro.build/en/guides/environment-variables/#intellisense-for-typescript) on behalf of the user. It's too basic and only handles static variables.
+
+## Dynamic variables using `Astro.env`
+
+The issue restrict dynamic variables usage inside `.astro` files (or endpoints `context`). It's common to be able to use it outside, eg. in `.ts`. That's why it uses an ALS.
+
+## Using zod
+
+Using zod in the public interface isn't great as `.env` files remain strings, so it requires more work to get right ([especially for booleans](https://env.t3.gg/docs/recipes#booleans)) and we only support a tiny subset of zod APIs.
+
+We've also considered using it under the hood. While it's not an issue for static variables (build time, not part of the bundle), it would increase the bundle for runtime usage significantly.
+
+## Providing fields helpers in a function
+
+Instead of exporting `envField` from `astro/config`, it has been considered to be provided as a function argument of `schema`:
+
+```ts
+export default defineConfig({
+  env: {
+    schema: (fields) => ({
+      FOO: fields.string()
+    })
+  }
+})
+```
+
+It's not great because:
+- It's not serializable
+- It can cause side-effects (if you call something before returning the object)
+- It makes merging in `updateConfig` (Integrations API) harder
+
+## Dedicated `env.ts`
+
+Having a dedicated entrypoint has been considered, mainly to play better with a zod based API:
+
+```ts
+import { defineEnv } from "astro/config"
+import { z } from "astro/zod"
+
+export default defineEnv({
+  FOO: z.string()
+})
+```
+
+It was too restrictive and caused 2 issues:
+- Users could forget to use the default export
+- Auto-complete could point to this file instead of virtual imports when typing `env`
+
+Note using such convention is still possible using the current proposal:
+
+```ts
+// env.ts
+import { envField } from "astro/config"
+
+export const envSchema = {
+  FOO: envField.static().public().string()
+} as const
```

Comment:
I think so because if you have a look at the detailed types above, public variables must start with `PUBLIC_` also at the type level. The `as const` allows users to catch the error earlier. An alternative here would be to provide a `defineEnvSchema` that allows to show errors directly.

That being said, it depends on what we want to promote: do we want to set this convention when there are too many variables or should we let people basically figure it out? (I'm in favor of such convention when it starts to grow significantly but not everyone was agreeing on the Stage 2 RFC)

Reply:
This is not relevant anymore, will update the proposal